Create optionally-available __int128 typedefs and use them for ARM64 definitions.

Gankra opened the pull request:
Potentially fixes #2524, see the comments in the patch for details.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Amanieu (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/libc/blob/master/CONTRIBUTING.md) for more information.

Gankra commented:
See https://github.com/rust-lang/rust/issues/54341#issuecomment-1062498329 for details on my investigation into this.

I will happily fill in more OS/arches if anyone can run https://github.com/Gankra/abi-checker on that target and get full passes. I only have access to aarch64 Android (phone) and macOS (coworker). I'm kinda blindly assuming iOS is the same as macOS here.

Gankra commented:
For context, properly defining these types is important for saving a crashing process' registers (i.e. when generating a breakpad-style minidump, which I am currently working on porting to Rust for Android).

bjorn3 commented on src/fixed_width_ints.rs:
```diff
@@ -18,3 +18,75 @@ pub type uint16_t = u16;
+    }
+}
+
+pub use self::supported_128::*;
```
nit: missing trailing newline

Gankra commented:
In the upstream issue nagisa notes that rust uniformly has this repr on all aarch64 targets -- but I'd still like to verify that the primary C compilers for those targets agree.

Amanieu commented on src/unix/linux_like/android/b64/aarch64/mod.rs:
```diff
@@ -55,6 +55,12 @@ s! {
         pub pstate: u64,
     }
+
+    pub struct user_fpsimd_struct {
```
Can you add this for Linux as well? It's the same struct, it's defined in the kernel headers.

Amanieu commented on src/fixed_width_ints.rs:
```diff
@@ -18,3 +18,70 @@ pub type uint16_t = u16;
+
+cfg_if! {
+    if #[cfg(all(target_arch = "aarch64",
+        any(target_os = "android", target_os="macos", target_os = "ios")))] {
```
I would feel comfortable just doing `not(windows)` here. That way all non-Windows platforms automatically get support and this is confirmed by the static assert.

Amanieu reviewed (commented):
Great work! I think this was sorely needed.

Gankra commented:
So since putting up this PR I spent a bit more time digging into what can be done to polyfill this type on x64 (since SysV actually passes structs the same as ints, in general). Unfortunately on x64 (linux, at least) clang and gcc actually disagree on the pass-on-stack convention for `__int128`:

https://twitter.com/Gankra_/status/1502320041315946499

x64 SysV fails to specify what "pushing" a value actually means and so clang packs __int128 while gcc properly aligns it. This can introduce padding and differences and completely desynchronize the offsets to args.

In discussing this on Twitter some Apple engineers also pointed out that clang actually has macos-specific ABI differences for backcompat with itself (details vague, no documentation for this in general). Similarly a coworker noted that stuff like this is a problem for clang plugins when the plugin and compiler weren't both consistently built by clang or gcc.

This now has me increasingly paranoid about claiming u128/i128 are ABI-compatible on ARM64 without any way for me to properly compare gcc and clang both targetting these platforms. And also makes me less inclined to willy-nilly declare that Rust agrees with C compilers for "all non-windows OSes". Even asserting they agree on size and align isn't sufficient because the compilers can still disagree on how the value is pushed to the stack. :(

I'm not sure what libc's policy is on divergent impls for one platform, and how to declare one "canonical" over the others.

Unfortunately I can't really test this stuff out better -- the "gcc" on my phone's termux is just an alias for clang, and similarly I expect that will be the case for m1 macbooks.

------------

If anyone has better access to arm64 platforms and can setup clang/gcc properly, they can test this out by checking out:

https://github.com/Gankra/abi-checker

and running

```
cargo run -- --tests ui128
```

This will compare CC to rustc and CC to itself. That is a good start for rust matching the "canonical" implementation. 

To exhaustively try out all the combos of clang+gcc+rust to check if the ABI is actually consistent between all impls, you can do:

```
cargo run -- --tests ui128 --pairs clang_calls_gcc gcc_calls_clang rustc_calls_gcc gcc_calls_rustc rustc_calls_clang clang_calls_rustc clang_calls_clang gcc_calls_gcc rustc_calls_rustc
```

------------

Otherwise you can manually do something like [this godbolt](https://godbolt.org/z/Prx8MTjrP).

Amanieu commented:
I ran your ABI checker on AArch64 Linux, all the tests pass:
```
Final Results:
ui128::c::clang_calls_gcc        all 226/226 passed
ui128::c::gcc_calls_clang        all 226/226 passed
ui128::c::rustc_calls_gcc        all 226/226 passed
ui128::c::gcc_calls_rustc        all 226/226 passed
ui128::c::rustc_calls_clang      all 226/226 passed
ui128::c::clang_calls_rustc      all 226/226 passed
ui128::c::clang_calls_clang      all 226/226 passed
ui128::c::gcc_calls_gcc          all 226/226 passed
ui128::c::rustc_calls_rustc      all 226/226 passed
ui128::cdecl::rustc_calls_rustc  all 226/226 passed
ui128::stdcall::rustc_calls_rustc all 226/226 passed
ui128::fastcall::rustc_calls_rustc all 226/226 passed

2712 passed, 0 failed, 0 completely failed, 33 skipped
```

Gankra commented:
Oh nice!! that helps my confidence with going forward with this a lot. I'll clean this up on monday.

Gankra commented:
Also lmao @ rustc actually managing to do *something* on linux when you ask it to fastcall/stdcall with itself, I hadn't actually every tested that.

Gankra commented:
Applied review requests and squashed.

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 3741d52dcba0d782cd27b1680e9d3dd481f33470 has been approved by `Amanieu`

<!-- @bors r=Amanieu 3741d52dcba0d782cd27b1680e9d3dd481f33470 -->
<!-- homu: {"type":"Approved","sha":"3741d52dcba0d782cd27b1680e9d3dd481f33470","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit 3741d52dcba0d782cd27b1680e9d3dd481f33470 with merge ca04d0981e9f44b4cad00b1a21850852895f2ec0...
<!-- homu: {"type":"BuildStarted","head_sha":"3741d52dcba0d782cd27b1680e9d3dd481f33470","merge_sha":"ca04d0981e9f44b4cad00b1a21850852895f2ec0"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5584859325?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5584859325?check_suite_focus=true","builder_name":"checks-actions"} -->

Amanieu commented on src/fixed_width_ints.rs:
```diff
@@ -18,3 +18,74 @@ pub type uint16_t = u16;
+        /// is actually *passed* on the stack (clang underaligns it for
+        /// the same reason that rustc *never* properly aligns it).
+        const _ASSERT_128_COMPAT: () = {
+            assert!(std::mem::size_of::<__int128>() == SIZE_128);
```
This should use `core` instead of `std`.

Amanieu commented on src/unix/linux_like/android/b64/aarch64/mod.rs:
```diff
@@ -55,6 +55,12 @@ s! {
     }
+
+    pub struct user_fpsimd_struct {
+        pub vregs: [__uint128_t; 32],
```
This type needs to be imported:
```suggestion
        pub vregs: [::__uint128_t; 32],
```

bors commented:
:pushpin: Commit c0d7485876f65785df4877b1db74e701dfa8a879 has been approved by `Amanieu`

<!-- @bors r=Amanieu c0d7485876f65785df4877b1db74e701dfa8a879 -->
<!-- homu: {"type":"Approved","sha":"c0d7485876f65785df4877b1db74e701dfa8a879","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit c0d7485876f65785df4877b1db74e701dfa8a879 with merge cda106ab42245f3936cdbde322785c2af2d74391...
<!-- homu: {"type":"BuildStarted","head_sha":"c0d7485876f65785df4877b1db74e701dfa8a879","merge_sha":"cda106ab42245f3936cdbde322785c2af2d74391"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5587984061?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5587984061?check_suite_focus=true","builder_name":"checks-actions"} -->

Gankra commented:
```
error: constant is never used: `SIZE_128`
  --> src/fixed_width_ints.rs:65:9
   |
65 |         const SIZE_128: usize = 16;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> src/lib.rs:18:38
   |
18 | #![cfg_attr(libc_deny_warnings, deny(warnings))]
   |                                      ^^^^^^^^
   = note: `#[deny(dead_code)]` implied by `#[deny(warnings)]`

error: constant is never used: `ALIGN_128`
  --> src/fixed_width_ints.rs:66:9
   |
66 |         const ALIGN_128: usize = 16;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `libc` due to 2 previous errors
```

??????? how ???????

bjorn3 commented:
Maybe `const _ASSERT_128_COMPAT: () = { ... }` doesn't mark them as used because it isn't used itself? Would `const _: () = { ... }` work?

Gankra commented:
It looks like prefixing them with _'s is the best approach.

pushed fix

bors commented:
:pushpin: Commit 80b951d67ab995da61b18588ba37d56dd3351f6e has been approved by `Amanieu`

<!-- @bors r=Amanieu 80b951d67ab995da61b18588ba37d56dd3351f6e -->
<!-- homu: {"type":"Approved","sha":"80b951d67ab995da61b18588ba37d56dd3351f6e","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit 80b951d67ab995da61b18588ba37d56dd3351f6e with merge c048d5be95f68c1b5f328396f566b8c2621a92b4...
<!-- homu: {"type":"BuildStarted","head_sha":"80b951d67ab995da61b18588ba37d56dd3351f6e","merge_sha":"c048d5be95f68c1b5f328396f566b8c2621a92b4"} -->

Gankra commented:
uhhh looks like the last remaining issue is... rust 1.13 support(????)

not sure what to do about that...

bjorn3 commented:
`proc_macro::is_available()` was stabilized on rustc 1.57 (https://github.com/rust-lang/rust/commit/09dd213cd2253015b80605e374d624fcf3c9aea0). CI uses a nightly from a couple of weeks before the moment it was stabilized, but because it is still 1.57, proc_macro2 thinks it was stabilized and tries to use it. 2021-10-11 is the first nightly with `is_available()` stabilized I think. You can either upgrade to it or pin proc_macro2 to 1.0.35 which doesn't use `is_available()` yet.

bjorn3 commented:
For the 1.13 issue you could conditionally compile it. The build script already detects a couple of rustc versions.

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 957e97b91d2f8e572853b49ccc8b246a874313eb has been approved by `Amanieu`

<!-- @bors r=Amanieu 957e97b91d2f8e572853b49ccc8b246a874313eb -->
<!-- homu: {"type":"Approved","sha":"957e97b91d2f8e572853b49ccc8b246a874313eb","approver":"Amanieu"} -->

Gankra commented:
I *think* I added the cfgs?

Since these defs are conditional I added a fallback for the old macos definition. That seems like the right thing to do?

bors commented:
:hourglass: Testing commit 957e97b91d2f8e572853b49ccc8b246a874313eb with merge 7bf5e77e2a66a24b50a30122c3361a66caf30f89...
<!-- homu: {"type":"BuildStarted","head_sha":"957e97b91d2f8e572853b49ccc8b246a874313eb","merge_sha":"7bf5e77e2a66a24b50a30122c3361a66caf30f89"} -->

JohnTitor commented:
> i128 was added in Rust 1.26, you might want to add a check for that in build.rs.
> 
> For the unreachable in garando_syntax I have no idea. cc @JohnTitor
> 
> ```
> thread 'main' panicked at 'internal error: entered unreachable code', /cargo/registry/src/github.com-1ecc6299db9ec823/garando_syntax-0.1.0/src/ext/expand.rs:337:21
> ```

I don't have enough time to investigate, but the code is quite old and I suspect that a newer thing here triggers that panic.
FWIW here's the repo: https://github.com/JohnTitor/garando

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5591589769?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5591589769?check_suite_focus=true","builder_name":"checks-actions"} -->

Gankra commented:
sorry i am unfortunately stuck in the situation of compile-by-ci here because I don't have direct access to these platforms on my dev machine.

* macos freaked out from the way i cfg'd, went with cfg_if with the s!'s being internal
* musl aarch64 complained it didn't have fpsimd; added it since it's a kernel API so it should be portable?

Gankra commented:
arghh nope macros bad style now 😭

Gankra commented:
Is there a "right" way to express what I'm doing with macos here? Should I just drop updating the macos def? Or not make it backcompat?

Amanieu commented:
I would just drop the type on old rustc which don't support i128. The type definition is incorrect in any case.

Amanieu commented:
@bors delegate=Gankra

bors commented:
:v: @Gankra can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"Amanieu","delegate":"Gankra"} -->

Gankra commented:
@bors try

Gankra commented:
@bors try

bors commented:
:hourglass: Trying commit 49400c79f7071c1cfa809684f2b9af27f9b83b4a with merge ab286d503d1b2a79db3cfba496bfa1881eaea521...
<!-- homu: {"type":"TryBuildStarted","head_sha":"49400c79f7071c1cfa809684f2b9af27f9b83b4a","merge_sha":"ab286d503d1b2a79db3cfba496bfa1881eaea521"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5593134557?check_suite_focus=true)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5593134557?check_suite_focus=true","builder_name":"checks-actions"} -->

Amanieu commented:
You can ignore the failure on sparc, it's spurious, just retry the build.

Gankra commented:
@bors try

bors commented:
:hourglass: Trying commit e07de0ee9e264b262c90ead5e049ff05d13d8ab2 with merge 3e9c0c0a3a53e139935c6c143731293fd85080fb...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e07de0ee9e264b262c90ead5e049ff05d13d8ab2","merge_sha":"3e9c0c0a3a53e139935c6c143731293fd85080fb"} -->

Gankra commented:
* const asserts only exist as of 1.57 so I added another cfg for that

* had to make a bunch more types optional behind int128 support in macos since the type is transitively used by several other ones.

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5593443992?check_suite_focus=true)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5593443992?check_suite_focus=true","builder_name":"checks-actions"} -->

Gankra commented:
@bors try

bors commented:
:hourglass: Trying commit 127542f58cd2e2e370062a5ff434c7c8d1749d68 with merge 1577ee48b59b1c08ebdfa1787a7f172da6891219...
<!-- homu: {"type":"TryBuildStarted","head_sha":"127542f58cd2e2e370062a5ff434c7c8d1749d68","merge_sha":"1577ee48b59b1c08ebdfa1787a7f172da6891219"} -->

Gankra commented:
trying to instead cfg the fields and keep the struct around, since the fallout from removing it seems kinda nasty

Gankra commented:
ok!!! this one actually seems to work and satisfy the nightmare cfg macro rube-goldberg machine (remaining failures are the pre-existing ones in the repo itself).

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5594281320?check_suite_focus=true)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5594281320?check_suite_focus=true","builder_name":"checks-actions"} -->

Gankra commented:
@bors r+

bors commented:
:pushpin: Commit 127542f58cd2e2e370062a5ff434c7c8d1749d68 has been approved by `Gankra`

<!-- @bors r=Gankra 127542f58cd2e2e370062a5ff434c7c8d1749d68 -->
<!-- homu: {"type":"Approved","sha":"127542f58cd2e2e370062a5ff434c7c8d1749d68","approver":"Gankra"} -->

bors commented:
:hourglass: Testing commit 127542f58cd2e2e370062a5ff434c7c8d1749d68 with merge aeb1a4e0b8b793db36065e40d0bab3118b6532b1...
<!-- homu: {"type":"BuildStarted","head_sha":"127542f58cd2e2e370062a5ff434c7c8d1749d68","merge_sha":"aeb1a4e0b8b793db36065e40d0bab3118b6532b1"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5604707523?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5604707523?check_suite_focus=true","builder_name":"checks-actions"} -->

Gankra commented:
ok coming back to this with a fresh mind and the design of libc i think now has congealed into a coherent form for me. Hopefully this works now...

Gankra commented:
@bors try

bors commented:
:hourglass: Trying commit adfa38507c1aa4adce7010a58825af35a1f5fb99 with merge 245c710f19b7fa29a1cad6bf9a3afc5000b69de0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"adfa38507c1aa4adce7010a58825af35a1f5fb99","merge_sha":"245c710f19b7fa29a1cad6bf9a3afc5000b69de0"} -->

Gankra commented:
@bors r+

bors commented:
:pushpin: Commit adfa38507c1aa4adce7010a58825af35a1f5fb99 has been approved by `Gankra`

<!-- @bors r=Gankra adfa38507c1aa4adce7010a58825af35a1f5fb99 -->
<!-- homu: {"type":"Approved","sha":"adfa38507c1aa4adce7010a58825af35a1f5fb99","approver":"Gankra"} -->

bors commented:
:hourglass: Testing commit adfa38507c1aa4adce7010a58825af35a1f5fb99 with merge f00ea005769ba76610002b3652a5c6272ddc3633...
<!-- homu: {"type":"BuildStarted","head_sha":"adfa38507c1aa4adce7010a58825af35a1f5fb99","merge_sha":"f00ea005769ba76610002b3652a5c6272ddc3633"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/libc/runs/5669045250?check_suite_focus=true), [checks-cirrus-freebsd-11](https://cirrus-ci.com/task/5946966281551872), [checks-cirrus-freebsd-12](https://cirrus-ci.com/task/5384016328130560), [checks-cirrus-freebsd-13](https://cirrus-ci.com/task/6509916234973184)
Approved by: Gankra
Pushing f00ea005769ba76610002b3652a5c6272ddc3633 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Gankra","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/libc/runs/5669045250?check_suite_focus=true","checks-cirrus-freebsd-11":"https://cirrus-ci.com/task/5946966281551872","checks-cirrus-freebsd-12":"https://cirrus-ci.com/task/5384016328130560","checks-cirrus-freebsd-13":"https://cirrus-ci.com/task/6509916234973184"},"merge_sha":"f00ea005769ba76610002b3652a5c6272ddc3633"} -->
